File: MiniDB_Projects/Team_LockAndRoll/src/parser.cpp
#include "parser.h"

#include <algorithm>
#include <cctype>
#include <unordered_set>

namespace minidb {

namespace {
const std::unordered_set<std::string> kKeywords = {
    "CREATE", "TABLE",  "INSERT",  "INTO",   "VALUES", "SELECT", "FROM",  "WHERE",
    "JOIN",   "INNER",  "ON",      "AND",    "OR",     "NOT",    "DELETE","BEGIN",
    "COMMIT", "ROLLBACK","ABORT",  "GROUP",  "BY",     "ORDER",  "ASC",   "DESC",
    "PRIMARY","KEY",    "AS",      "TRUE",   "FALSE",  "NULL",   "INTEGER","VARCHAR",
    "BOOLEAN","INT",    "TEXT",    "LIMIT",  "COUNT",  "SUM",    "AVG",   "MIN",
    "MAX"};

std::string upper(std::string s) {
  for (char& c : s) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
  return s;
}
bool is_agg(const std::string& kw) {
  return kw == "COUNT" || kw == "SUM" || kw == "AVG" || kw == "MIN" || kw == "MAX";
}

// turns overflow into a clean DBException instead of letting std::out_of_range escape
int64_t to_int64(const std::string& text) {
  try {
    return std::stoll(text);
  } catch (const std::out_of_range&) {
    throw DBException("integer literal out of range: " + text);
  } catch (const std::invalid_argument&) {
    throw DBException("invalid integer literal: " + text);
  }
}
}  // namespace

std::vector<Token> Parser::lex(const std::string& sql) {
  std::vector<Token> toks;
  size_t i = 0, n = sql.size();
  while (i < n) {
    char c = sql[i];
    if (std::isspace(static_cast<unsigned char>(c))) { i++; continue; }
    // -- line comment
    if (c == '-' && i + 1 < n && sql[i + 1] == '-') {
      while (i < n && sql[i] != '\n') i++;
      continue;
    }
    if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
      size_t j = i;
      while (j < n && (std::isalnum(static_cast<unsigned char>(sql[j])) || sql[j] == '_')) j++;
      std::string word = sql.substr(i, j - i);
      std::string up = upper(word);
      if (kKeywords.count(up))
        toks.push_back({Token::Type::Keyword, up});
      else
        toks.push_back({Token::Type::Ident, word});
      i = j;
      continue;
    }
    if (std::isdigit(static_cast<unsigned char>(c))) {
      size_t j = i;
      while (j < n && std::isdigit(static_cast<unsigned char>(sql[j]))) j++;
      toks.push_back({Token::Type::Number, sql.substr(i, j - i)});
      i = j;
      continue;
    }
    if (c == '\'') {  // '' is an escaped quote
      std::string s;
      i++;
      bool closed = false;
      while (i < n) {
        if (sql[i] == '\'') {
          if (i + 1 < n && sql[i + 1] == '\'') { s.push_back('\''); i += 2; continue; }
          i++;
          closed = true;
          break;
        }
        s.push_back(sql[i++]);
      }
      if (!closed) throw DBException("unterminated string literal");
      toks.push_back({Token::Type::String, s});
      continue;
    }
    auto two = (i + 1 < n) ? sql.substr(i, 2) : std::string();
    if (two == "<=" || two == ">=" || two == "!=" || two == "<>") {
      toks.push_back({Token::Type::Punct, two});
      i += 2;
      continue;
    }
    if (std::string("()*,.;=<>+-/").find(c) != std::string::npos) {
      toks.push_back({Token::Type::Punct, std::string(1, c)});
      i++;
      continue;
    }
    throw DBException(std::string("unexpected character '") + c + "' in SQL");
  }
  toks.push_back({Token::Type::End, ""});
  return toks;
}

void Parser::fail(const std::string& msg) const {
  const Token& t = toks_[pos_];
  throw DBException("parse error near '" + t.text + "': " + msg);
}
bool Parser::accept_kw(const std::string& kw) {
  if (is_kw(kw)) { pos_++; return true; }
  return false;
}
bool Parser::accept_punct(const std::string& p) {
  if (is_punct(p)) { pos_++; return true; }
  return false;
}
void Parser::expect_kw(const std::string& kw) {
  if (!accept_kw(kw)) fail("expected keyword " + kw);
}
void Parser::expect_punct(const std::string& p) {
  if (!accept_punct(p)) fail("expected '" + p + "'");
}
std::string Parser::expect_ident() {
  if (peek().type != Token::Type::Ident) fail("expected identifier");
  return next().text;
}

Statement Parser::parse(const std::string& sql) {
  Parser p(lex(sql));
  Statement s = p.parse_statement();
  p.accept_punct(";");
  if (p.peek().type != Token::Type::End) p.fail("unexpected trailing tokens");
  return s;
}

Statement Parser::parse_statement() {
  if (is_kw("CREATE")) return parse_create();
  if (is_kw("INSERT")) return parse_insert();
  if (is_kw("SELECT")) return parse_select();
  if (is_kw("DELETE")) return parse_delete();
  if (accept_kw("BEGIN")) return TxnStmt{TxnCmd::Begin};
  if (accept_kw("COMMIT")) return TxnStmt{TxnCmd::Commit};
  if (accept_kw("ROLLBACK") || accept_kw("ABORT")) return TxnStmt{TxnCmd::Rollback};
  fail("unrecognized statement");
}

CreateTableStmt Parser::parse_create() {
  expect_kw("CREATE");
  expect_kw("TABLE");
  CreateTableStmt s;
  s.table = expect_ident();
  expect_punct("(");
  do {
    ColumnDef col;
    col.name = expect_ident();
    if (accept_kw("INTEGER") || accept_kw("INT")) col.type = TypeId::INTEGER;
    else if (accept_kw("VARCHAR") || accept_kw("TEXT")) col.type = TypeId::VARCHAR;
    else if (accept_kw("BOOLEAN")) col.type = TypeId::BOOLEAN;
    else fail("expected a column type");
    // optional VARCHAR(n) length: accepted and ignored
    if (accept_punct("(")) { next(); expect_punct(")"); }
    if (accept_kw("PRIMARY")) { expect_kw("KEY"); col.primary_key = true; }
    s.columns.push_back(std::move(col));
  } while (accept_punct(","));
  expect_punct(")");
  return s;
}

InsertStmt Parser::parse_insert() {
  expect_kw("INSERT");
  expect_kw("INTO");
  InsertStmt s;
  s.table = expect_ident();
  if (accept_punct("(")) {
    do { s.columns.push_back(expect_ident()); } while (accept_punct(","));
    expect_punct(")");
  }
  expect_kw("VALUES");
  do {
    expect_punct("(");
    std::vector<ExprPtr> row;
    do { row.push_back(parse_expr()); } while (accept_punct(","));
    expect_punct(")");
    s.rows.push_back(std::move(row));
  } while (accept_punct(","));
  return s;
}

SelectStmt Parser::parse_select() {
  expect_kw("SELECT");
  SelectStmt s;
  do {
    SelectItem item;
    if (accept_punct("*")) {
      item.star = true;
    } else {
      item.expr = parse_expr();
      if (accept_kw("AS")) item.alias = expect_ident();
    }
    s.items.push_back(std::move(item));
  } while (accept_punct(","));

  expect_kw("FROM");
  s.from.name = expect_ident();
  if (peek().type == Token::Type::Ident && !is_kw("WHERE")) s.from.alias = next().text;

  while (is_kw("JOIN") || is_kw("INNER")) {
    accept_kw("INNER");
    expect_kw("JOIN");
    JoinClause jc;
    jc.right.name = expect_ident();
    if (peek().type == Token::Type::Ident && !is_kw("ON")) jc.right.alias = next().text;
    expect_kw("ON");
    jc.on = parse_expr();
    s.joins.push_back(std::move(jc));
  }

  if (accept_kw("WHERE")) s.where = parse_expr();

  if (accept_kw("GROUP")) {
    expect_kw("BY");
    do { s.group_by.push_back(parse_primary()->col_name); } while (accept_punct(","));
  }

  if (accept_kw("ORDER")) {
    expect_kw("BY");
    do {
      OrderByItem ob;
      ob.col = parse_primary()->col_name;
      if (accept_kw("DESC")) ob.desc = true;
      else accept_kw("ASC");
      s.order_by.push_back(std::move(ob));
    } while (accept_punct(","));
  }

  if (accept_kw("LIMIT")) {
    s.has_limit = true;
    if (peek().type != Token::Type::Number) fail("expected a number after LIMIT");
    s.limit = to_int64(next().text);
  }
  return s;
}

DeleteStmt Parser::parse_delete() {
  expect_kw("DELETE");
  expect_kw("FROM");
  DeleteStmt s;
  s.table = expect_ident();
  if (accept_kw("WHERE")) s.where = parse_expr();
  return s;
}

ExprPtr Parser::parse_expr() { return parse_or(); }

ExprPtr Parser::parse_or() {
  ExprPtr l = parse_and();
  while (accept_kw("OR")) {
    auto e = std::make_shared<Expr>();
    e->kind = ExprKind::Binary;
    e->op = "OR";
    e->left = l;
    e->right = parse_and();
    l = e;
  }
  return l;
}

ExprPtr Parser::parse_and() {
  ExprPtr l = parse_cmp();
  while (accept_kw("AND")) {
    auto e = std::make_shared<Expr>();
    e->kind = ExprKind::Binary;
    e->op = "AND";
    e->left = l;
    e->right = parse_cmp();
    l = e;
  }
  return l;
}

ExprPtr Parser::parse_cmp() {
  ExprPtr l = parse_add();
  for (const char* op : {"=", "!=", "<>", "<", "<=", ">", ">="}) {
    if (is_punct(op)) {
      std::string o = next().text;
      if (o == "<>") o = "!=";
      auto e = std::make_shared<Expr>();
      e->kind = ExprKind::Binary;
      e->op = o;
      e->left = l;
      e->right = parse_add();
      return e;
    }
  }
  return l;
}

ExprPtr Parser::parse_add() {
  ExprPtr l = parse_mul();
  while (is_punct("+") || is_punct("-")) {
    std::string o = next().text;
    auto e = std::make_shared<Expr>();
    e->kind = ExprKind::Binary;
    e->op = o;
    e->left = l;
    e->right = parse_mul();
    l = e;
  }
  return l;
}

ExprPtr Parser::parse_mul() {
  ExprPtr l = parse_primary();
  while (is_punct("*") || is_punct("/")) {
    std::string o = next().text;
    auto e = std::make_shared<Expr>();
    e->kind = ExprKind::Binary;
    e->op = o;
    e->left = l;
    e->right = parse_primary();
    l = e;
  }
  return l;
}

ExprPtr Parser::parse_primary() {
  auto e = std::make_shared<Expr>();
  if (accept_punct("(")) {
    ExprPtr inner = parse_expr();
    expect_punct(")");
    return inner;
  }
  if (is_punct("-")) {
    next();
    e->kind = ExprKind::Unary;
    e->op = "-";
    e->left = parse_primary();
    return e;
  }
  if (peek().type == Token::Type::Number) {
    e->kind = ExprKind::IntLit;
    e->int_val = to_int64(next().text);
    return e;
  }
  if (peek().type == Token::Type::String) {
    e->kind = ExprKind::StrLit;
    e->str_val = next().text;
    return e;
  }
  if (accept_kw("TRUE")) { e->kind = ExprKind::BoolLit; e->bool_val = true; return e; }
  if (accept_kw("FALSE")) { e->kind = ExprKind::BoolLit; e->bool_val = false; return e; }
  if (accept_kw("NULL")) { e->kind = ExprKind::NullLit; return e; }

  if (peek().type == Token::Type::Keyword && is_agg(peek().text)) {
    e->kind = ExprKind::Agg;
    e->func = next().text;
    expect_punct("(");
    if (accept_punct("*")) e->star = true;
    else e->left = parse_expr();
    expect_punct(")");
    return e;
  }

  if (peek().type == Token::Type::Ident) {
    std::string name = next().text;
    if (accept_punct(".")) name += "." + expect_ident();
    e->kind = ExprKind::Column;
    e->col_name = name;
    return e;
  }
  fail("expected an expression");
}

}  // namespace minidb

File: MiniDB_Projects/Team_candyman/internal/sql/parser.go
package sql

import (
	"fmt"
	"strconv"

	"minidb/internal/types"
)

// Parse parses a single SQL statement (without a trailing semicolon).
func Parse(input string) (Statement, error) {
	toks, err := newLexer(input).tokens()
	if err != nil {
		return nil, err
	}
	p := &parser{toks: toks}
	stmt, err := p.parseStatement()
	if err != nil {
		return nil, err
	}
	if p.peek().kind != tEOF {
		return nil, fmt.Errorf("sql: unexpected token %q after statement", p.peek().text)
	}
	return stmt, nil
}

type parser struct {
	toks []token
	pos  int
}

func (p *parser) peek() token { return p.toks[p.pos] }
func (p *parser) advance() token {
	t := p.toks[p.pos]
	if p.pos < len(p.toks)-1 {
		p.pos++
	}
	return t
}
func (p *parser) acceptKeyword(kw string) bool {
	if t := p.peek(); t.kind == tKeyword && t.text == kw {
		p.advance()
		return true
	}
	return false
}
func (p *parser) expectKeyword(kw string) error {
	if !p.acceptKeyword(kw) {
		return fmt.Errorf("sql: expected %s, got %q", kw, p.peek().text)
	}
	return nil
}
func (p *parser) acceptPunct(s string) bool {
	if t := p.peek(); t.kind == tPunct && t.text == s {
		p.advance()
		return true
	}
	return false
}
func (p *parser) expectPunct(s string) error {
	if !p.acceptPunct(s) {
		return fmt.Errorf("sql: expected %q, got %q", s, p.peek().text)
	}
	return nil
}
func (p *parser) expectIdent() (string, error) {
	t := p.peek()
	if t.kind != tIdent {
		return "", fmt.Errorf("sql: expected identifier, got %q", t.text)
	}
	p.advance()
	return t.text, nil
}

func (p *parser) parseStatement() (Statement, error) {
	t := p.peek()
	if t.kind != tKeyword {
		return nil, fmt.Errorf("sql: expected a statement keyword, got %q", t.text)
	}
	switch t.text {
	case "CREATE":
		return p.parseCreate()
	case "INSERT":
		return p.parseInsert()
	case "DELETE":
		return p.parseDelete()
	case "SELECT", "EXPLAIN":
		return p.parseSelect()
	case "BEGIN":
		p.advance()
		return &Begin{}, nil
	case "COMMIT":
		p.advance()
		return &Commit{}, nil
	case "ROLLBACK":
		p.advance()
		return &Rollback{}, nil
	default:
		return nil, fmt.Errorf("sql: unsupported statement %q", t.text)
	}
}

func (p *parser) parseCreate() (Statement, error) {
	p.advance() // CREATE
	if err := p.expectKeyword("TABLE"); err != nil {
		return nil, err
	}
	name, err := p.expectIdent()
	if err != nil {
		return nil, err
	}
	if err := p.expectPunct("("); err != nil {
		return nil, err
	}
	ct := &CreateTable{Table: name}
	seenPK := false
	for {
		col, err := p.expectIdent()
		if err != nil {
			return nil, err
		}
		tk := p.peek()
		if tk.kind != tKeyword {
			return nil, fmt.Errorf("sql: expected column type, got %q", tk.text)
		}
		colType, err := types.ParseType(tk.text)
		if err != nil {
			return nil, err
		}
		p.advance()
		def := ColumnDef{Name: col, Type: colType}
		if p.acceptKeyword("PRIMARY") {
			if err := p.expectKeyword("KEY"); err != nil {
				return nil, err
			}
			def.PrimaryKey = true
			seenPK = true
		}
		ct.Columns = append(ct.Columns, def)
		if p.acceptPunct(",") {
			continue
		}
		break
	}
	if err := p.expectPunct(")"); err != nil {
		return nil, err
	}
	if !seenPK {
		return nil, fmt.Errorf("sql: CREATE TABLE %s requires a PRIMARY KEY column", name)
	}
	return ct, nil
}

func (p *parser) parseInsert() (Statement, error) {
	p.advance() // INSERT
	if err := p.expectKeyword("INTO"); err != nil {
		return nil, err
	}
	name, err := p.expectIdent()
	if err != nil {
		return nil, err
	}
	ins := &Insert{Table: name}
	if p.acceptPunct("(") {
		for {
			col, err := p.expectIdent()
			if err != nil {
				return nil, err
			}
			ins.Columns = append(ins.Columns, col)
			if p.acceptPunct(",") {
				continue
			}
			break
		}
		if err := p.expectPunct(")"); err != nil {
			return nil, err
		}
	}
	if err := p.expectKeyword("VALUES"); err != nil {
		return nil, err
	}
	for {
		if err := p.expectPunct("("); err != nil {
			return nil, err
		}
		var row []Expr
		for {
			e, err := p.parsePrimary()
			if err != nil {
				return nil, err
			}
			row = append(row, e)
			if p.acceptPunct(",") {
				continue
			}
			break
		}
		if err := p.expectPunct(")"); err != nil {
			return nil, err
		}
		ins.Rows = append(ins.Rows, row)
		if p.acceptPunct(",") {
			continue
		}
		break
	}
	return ins, nil
}

func (p *parser) parseDelete() (Statement, error) {
	p.advance() // DELETE
	if err := p.expectKeyword("FROM"); err != nil {
		return nil, err
	}
	name, err := p.expectIdent()
	if err != nil {
		return nil, err
	}
	del := &Delete{Table: name}
	if p.acceptKeyword("WHERE") {
		w, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		del.Where = w
	}
	return del, nil
}

func (p *parser) parseSelect() (Statement, error) {
	sel := &Select{}
	if p.acceptKeyword("EXPLAIN") {
		sel.Explain = true
	}
	if err := p.expectKeyword("SELECT"); err != nil {
		return nil, err
	}
	if p.acceptPunct("*") {
		sel.Star = true
	} else {
		for {
			e, err := p.parseExpr()
			if err != nil {
				return nil, err
			}
			item := SelectItem{Expr: e}
			if p.acceptKeyword("AS") {
				alias, err := p.expectIdent()
				if err != nil {
					return nil, err
				}
				item.Alias = alias
			}
			sel.Items = append(sel.Items, item)
			if p.acceptPunct(",") {
				continue
			}
			break
		}
	}
	if err := p.expectKeyword("FROM"); err != nil {
		return nil, err
	}
	// table list with optional JOIN ... ON predicates
	tr, err := p.parseTableRef()
	if err != nil {
		return nil, err
	}
	sel.Tables = append(sel.Tables, tr)
	for {
		if p.acceptPunct(",") {
			tr, err := p.parseTableRef()
			if err != nil {
				return nil, err
			}
			sel.Tables = append(sel.Tables, tr)
			continue
		}
		if p.acceptKeyword("JOIN") {
			tr, err := p.parseTableRef()
			if err != nil {
				return nil, err
			}
			sel.Tables = append(sel.Tables, tr)
			if err := p.expectKeyword("ON"); err != nil {
				return nil, err
			}
			cond, err := p.parseExpr()
			if err != nil {
				return nil, err
			}
			sel.Predicates = append(sel.Predicates, cond)
			continue
		}
		break
	}
	if p.acceptKeyword("WHERE") {
		w, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		sel.Predicates = append(sel.Predicates, splitAnd(w)...)
	}
	if p.acceptKeyword("GROUP") {
		if err := p.expectKeyword("BY"); err != nil {
			return nil, err
		}
		for {
			e, err := p.parsePrimary()
			if err != nil {
				return nil, err
			}
			sel.GroupBy = append(sel.GroupBy, e)
			if p.acceptPunct(",") {
				continue
			}
			break
		}
	}
	return sel, nil
}

func (p *parser) parseTableRef() (TableRef, error) {
	name, err := p.expectIdent()
	if err != nil {
		return TableRef{}, err
	}
	ref := TableRef{Name: name, Alias: name}
	if p.acceptKeyword("AS") {
		alias, err := p.expectIdent()
		if err != nil {
			return TableRef{}, err
		}
		ref.Alias = alias
	} else if t := p.peek(); t.kind == tIdent {
		ref.Alias = t.text
		p.advance()
	}
	return ref, nil
}

// parseExpr parses AND-connected comparisons.
func (p *parser) parseExpr() (Expr, error) {
	left, err := p.parseComparison()
	if err != nil {
		return nil, err
	}
	for p.acceptKeyword("AND") {
		right, err := p.parseComparison()
		if err != nil {
			return nil, err
		}
		left = &BinaryExpr{Op: "AND", Left: left, Right: right}
	}
	return left, nil
}

func (p *parser) parseComparison() (Expr, error) {
	left, err := p.parsePrimary()
	if err != nil {
		return nil, err
	}
	if t := p.peek(); t.kind == tOp {
		p.advance()
		right, err := p.parsePrimary()
		if err != nil {
			return nil, err
		}
		op := t.text
		if op == "<>" {
			op = "!="
		}
		return &BinaryExpr{Op: op, Left: left, Right: right}, nil
	}
	return left, nil
}

func (p *parser) parsePrimary() (Expr, error) {
	t := p.peek()
	switch t.kind {
	case tNumber:
		p.advance()
		n, err := strconv.ParseInt(t.text, 10, 64)
		if err != nil {
			return nil, fmt.Errorf("sql: bad integer %q", t.text)
		}
		return &Literal{Value: types.NewInt(n)}, nil
	case tString:
		p.advance()
		return &Literal{Value: types.NewText(t.text)}, nil
	case tKeyword:
		switch t.text {
		case "NULL":
			p.advance()
			return &Literal{Value: types.Value{Null: true}}, nil
		case "COUNT", "SUM", "AVG", "MIN", "MAX":
			return p.parseAgg()
		}
		return nil, fmt.Errorf("sql: unexpected keyword %q in expression", t.text)
	case tIdent:
		p.advance()
		ref := &ColumnRef{Name: t.text}
		if p.acceptPunct(".") {
			col, err := p.expectIdent()
			if err != nil {
				return nil, err
			}
			ref.Table = t.text
			ref.Name = col
		}
		return ref, nil
	case tPunct:
		if t.text == "(" {
			p.advance()
			e, err := p.parseExpr()
			if err != nil {
				return nil, err
			}
			if err := p.expectPunct(")"); err != nil {
				return nil, err
			}
			return e, nil
		}
	}
	return nil, fmt.Errorf("sql: unexpected token %q in expression", t.text)
}

func (p *parser) parseAgg() (Expr, error) {
	fn := p.advance().text // COUNT/SUM/...
	if err := p.expectPunct("("); err != nil {
		return nil, err
	}
	agg := &AggCall{Func: fn}
	if p.acceptPunct("*") {
		agg.Star = true
	} else {
		arg, err := p.parseExpr()
		if err != nil {
			return nil, err
		}
		agg.Arg = arg
	}
	if err := p.expectPunct(")"); err != nil {
		return nil, err
	}
	return agg, nil
}

// splitAnd flattens an AND tree into a list of conjuncts.
func splitAnd(e Expr) []Expr {
	be, ok := e.(*BinaryExpr)
	if ok && be.Op == "AND" {
		return append(splitAnd(be.Left), splitAnd(be.Right)...)
	}
	return []Expr{e}
}
